cpp/CMakeLists.txt

```diff
@@ -938,8 +943,60 @@ set(ARROW_STATIC_PRIVATE_LINK_LIBS
 if (NOT MSVC)
   set(ARROW_LINK_LIBS
     ${ARROW_LINK_LIBS}
-    ${CMAKE_DL_LIBS}
-    pthread)
+    ${CMAKE_DL_LIBS})
+endif()
+
+if (ARROW_JEMALLOC)
+  add_definitions(-DARROW_JEMALLOC)
+  # In the case that jemalloc is only available as a shared library also use it to
+  # link it in the static requirements. In contrast to other libraries we try in
+  # most cases to use the system provided version of jemalloc to better align with
+  # other potential users of jemalloc.
+  if (JEMALLOC_STATIC_LIB)
+    set(ARROW_JEMALLOC_STATIC_LINK_LIBS
+      jemalloc_static
+      )
+  else()
+    set(ARROW_JEMALLOC_STATIC_LINK_LIBS
+      jemalloc_shared
+      )
+  endif()
+
+  if (NOT APPLE)
+      set(ARROW_JEMALLOC_STATIC_LINK_LIBS ${ARROW_JEMALLOC_STATIC_LINK_LIBS} ${PTHREAD_LIBRARY})
+  endif()
+
+  if (ARROW_JEMALLOC_USE_SHARED)
+    set(ARROW_JEMALLOC_SHARED_LINK_LIBS
+      jemalloc_shared
+      )
+  else()
+    if (CMAKE_COMPILER_IS_GNUCXX)
```

Comment:
I'm not sure about that. I think it mostly impacts glibc versions that are installed on CentOS 5 and Ubuntu 12.04. On these ancient systems, I don't think we could build arrow using clang, thus I would not further investigate if this is specific to glibc or glibc+gcc.